### PR TITLE
TST: fix for RTD failure with new matplotlib version

### LINKED ISSUE #13209: RTD failure due to plots involving convolution Kernels
### Description

There are RTD failures due to a missing `ndim` in latest PRs, see:

- https://readthedocs.org/projects/astropy/builds/16812769/
- https://readthedocs.org/projects/astropy/builds/16812180/

I believe this may have to do with the latest version of `matplotlib` (https://github.com/matplotlib/matplotlib/releases/tag/v3.5.2) interacting poorly with how plots for `Box1DKernel`, `Gaussian1DKernel`, `RickerWavelet1DKernel`, and `Trapezoid1DKernel` are coded.

Note that a simple attempt at fixing this was tried in #13205; however, I think the fix will be more involved than this.

This is related to matplotlib/matplotlib#22973

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

RTD failures due to a missing `ndim` in latest PRs, see:

- https://readthedocs.org/projects/astropy/builds/16812769/
- https://readthedocs.org/projects/astropy/builds/16812180/

Fix #13209

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/convolution/core.py [hunk 1]
```diff
@@ -243,6 +243,10 @@ def __init__(self, model=None, x_size=None, array=None, **kwargs):
 
         super().__init__(array)
 
+    @property
+    def ndim(self):
```
Comment (pllim) [inline astropy/convolution/core.py:247]: Do you think it is cleaner to return `self._array.ndim` in the base class, like `shape`?

Comment (mhvk) [reply astropy/convolution/core.py:247]: Yes, using `self._array.ndim` would seem cleaner. This is also used in the `dimension` property. I think it makes sense to go with it here too. More generally, I think it makes sense to add this property, but perhaps with a docstring that states it is for compatibility with `np.array`, something like """Kernal dimension. Alias to dimension, for compatibility with arrays.""" (unless we want to deprecate `dimension` in favour of `ndim`).

Comment (WilliamJamieson) [reply astropy/convolution/core.py:247]: This approach has turned into a dead-end for addressing the larger issue. It now fails with:
```python
/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/13205/lib/python3.9/site-packages/astropy/convolution/kernels.py:docstring of astropy.convolution.kernels.Box1DKernel:56: WARNING: Exception occurred in plotting astropy-convolution-Box1DKernel-1
 from /home/docs/checkouts/readthedocs.org/user_builds/astropy/checkouts/13205/docs/api/astropy.convolution.Box1DKernel.rst:
Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/13205/lib/python3.9/site-packages/matplotlib/sphinxext/plot_directive.py", line 517, in _run_code
    exec(code, ns)
  File "<string>", line 4, in <module>
  File "/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/13205/lib/python3.9/site-packages/matplotlib/pyplot.py", line 2769, in plot
    return gca().plot(
  File "/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/13205/lib/python3.9/site-packages/matplotlib/axes/_axes.py", line 1632, in plot
    lines = [*self._get_lines(*args, data=data, **kwargs)]
  File "/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/13205/lib/python3.9/site-packages/matplotlib/axes/_base.py", line 312, in __call__
    yield from self._plot_args(this, kwargs)
  File "/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/13205/lib/python3.9/site-packages/matplotlib/axes/_base.py", line 506, in _plot_args
    y = y[:, np.newaxis]
TypeError: 'Box1DKernel' object is not subscriptable
```

Any suggestions?

Comment (mhvk) [reply astropy/convolution/core.py:247]: So, clearly matplotlib is no longer coercing the input to arrays. Maybe then the best thing is to adjust the actual examples (i.e., use `kernel.array` or whatever it was), and add an entry to API changes or so. The only alternative is to behave even more like an array, but then the overrides of `__mul__` to mean convolution, etc., are becoming quite weird.

**Comment by pllim** [comment]
@WilliamJamieson , if you reported this upstream, please crosslink here. Thanks!

**Comment by WilliamJamieson** [comment]
> I am torn between merging this or pinning a `!=` for the offending matplotlib version in docs requirements. I am afraid we will forget to undo this when upstream gets fixed.

I guess we could leave #13209 open (maybe update the title to something better) until matplotlib fixes the issue.

**Comment by WilliamJamieson** [comment]
> > I guess we could leave #13209 open (maybe update the title to something better) until matplotlib fixes the issue.
> 
> Yes, but in the meantime, I'd suggest pinning a `!=` so our RTD runs again.

c822154 switches over to doing this.

**Comment by pllim** [comment]
Thanks! The diff LGTM but let's see what CI says. And I think this needs squashing?

**Review by pllim — APPROVED** [review]
Yay, RTD is green. Thanks!

**Comment by pllim** [comment]
@meeseeksdev backport to v5.0.x

**Comment by pllim** [comment]
@meeseeksdev backport to v5.1.x

**Pull request merged**